File: testdata/fixtures/java/CWE-94.java
package fixture;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.servlet.http.HttpServletRequest;
import org.springframework.expression.spel.standard.SpelExpressionParser;

class Cwe94 {

    void viaScriptEngine(HttpServletRequest request) throws Exception {
        String code = request.getParameter("code");
        ScriptEngine engine = new ScriptEngineManager().getEngineByName("javascript");
        // ruleid: java-code-injection
        engine.eval(code);
    }

    void viaScriptEngineAlt(HttpServletRequest request) throws Exception {
        String payload = request.getHeader("X-Code");
        ScriptEngine engine = new ScriptEngineManager().getEngineByName("groovy");
        // ruleid: java-code-injection
        engine.eval(payload);
    }

    void viaSpel(HttpServletRequest request) {
        String expr = request.getParameter("expr");
        SpelExpressionParser parser = new SpelExpressionParser();
        // ruleid: java-code-injection
        parser.parseExpression(expr).getValue();
    }

    void viaConcatenated(HttpServletRequest request) throws Exception {
        String code = request.getParameter("code");
        String full = "function test(){" + code + "}";
        ScriptEngine engine = new ScriptEngineManager().getEngineByName("js");
        // ruleid: java-code-injection
        engine.eval(full);
    }

    // ok: java-code-injection
    void safeConstant() throws Exception {
        ScriptEngine engine = new ScriptEngineManager().getEngineByName("js");
        engine.eval("1+1");
    }

    // ok: java-code-injection
    void safeSpelConstant() {
        SpelExpressionParser parser = new SpelExpressionParser();
        parser.parseExpression("'hello'.concat(' world')").getValue();
    }

    void safeFromConfig(String configured) throws Exception {
        ScriptEngine engine = new ScriptEngineManager().getEngineByName("js");
        // ok: java-code-injection
        engine.eval(configured);
    }
}

File: testdata/fixtures/scala/CWE-94.scala
// CWE-94: Code Injection via ToolBox and ScriptEngine
//
// Untrusted input passed to ToolBox.eval or ScriptEngine.eval is executed
// as Scala/Java code. The source is a request parameter.

import javax.servlet.http.HttpServletRequest
import scala.reflect.runtime.currentMirror
import scala.tools.reflect.ToolBox
import javax.script.{ScriptEngineManager, ScriptEngine}

object Cwe94CodeInjection {
  def viaToolbox(request: HttpServletRequest): Unit = {
    val code = request.getParameter("code")
    val tb = currentMirror.mkToolBox()
    // ruleid: scala-code-injection
    tb.eval(tb.parse(code))
  }

  def viaEngine(request: HttpServletRequest): Unit = {
    val expr = request.getParameter("expr")
    val engine: ScriptEngine = new ScriptEngineManager().getEngineByName("js")
    // ruleid: scala-code-injection
    engine.eval(expr)
  }

  def viaDirect(request: HttpServletRequest): Unit = {
    val input = request.getParameter("input")
    val engine = new ScriptEngineManager().getEngineByName("scala")
    // ruleid: scala-code-injection
    engine.eval(input)
  }

  def safe(): Unit = {
    // ok: scala-code-injection
    val tb = currentMirror.mkToolBox()
    tb.eval(tb.parse("1 + 1"))
    // ok: scala-code-injection
    val engine = new ScriptEngineManager().getEngineByName("js")
    engine.eval("1+1")
    // ok: scala-code-injection
    val code = "fixed"
    tb.eval(tb.parse(code))
  }
}
